### PR TITLE
Add implementations of `VerifiedSemigroup Nat` etc to contrib

### PR DESCRIPTION
**Pull request by atacratic**
New interfaces and implementations in contrib, building up the coverage of `Verified` implementations which provide proof terms to witness that the base instance obeys the appropriate laws.  Includes a back-compatibility breakage - demotion of the instance of `Semigroup Nat` into a named instance.

Example: `VerifiedSemigroup Nat` contains a witness for associativity of the operation defined in `Semigroup Nat`.

- New implementations of algebraic interfaces for numeric types, for example SemiGroup for multiplication on Nat (as well as addition).  ZZ now has the whole numeric tower covered up to `RingWithUnity` - ditto Nat where appropriate.

- All these instances now have `Verified` counterparts as well (libs/contrib/Interfaces/Verified.idr), for example the instance `[PlusNatSemiV] VerifiedSemigroup Nat` extends `[PlusNatSemi] Semigroup Nat` with a member that is a proof term `semigroupOpIsAssociative`.

- Some non-numeric `Verified` interfaces and instances are added too - for example `VerifiedFunctor Maybe`.  There are TODO placeholders for extending coverage further (`VerifiedFunctor List` for example).

- Back-compatibility break in contrib: in libs/contrib/Control/Algebra/NumericImplementations.idr, the `Semigroup Nat` instance is demoted into a named instance called `PlusNatSemi`.  This puts it on a par with the new `MultNatSemi`.  Similarly `Monoid Nat`, `SemiGroup ZZ` and `Monoid ZZ`.

### CODE REVIEW

**Comment by stephen-smith** [comment]
Can the `PlusNat*` instances be used for `Prelude.Nat.Additive` and the `MultiNat*` instances sued for `Prelude.Nat.Multiplicative`, too?

File: libs/contrib/Interfaces/Verified.idr [hunk 1]
```diff
@@ -40,49 +69,118 @@ interface (Monad m, VerifiedApplicative m) => VerifiedMonad (m : Type -> Type) w
...
+
+[MultZZMonoidV] VerifiedMonoid ZZ using MultZZSemiV, MultZZMonoid where
+  monoidNeutralIsNeutralL = multOneRightNeutralZ
+  monoidNeutralIsNeutralR = multOneLeftNeutralZ
 
 implementation VerifiedMonoid (List a) where
   monoidNeutralIsNeutralL = appendNilRightNeutral
   monoidNeutralIsNeutralR xs = Refl
 
+-- TODO Verified versions of...
```
Comment (ahmadsalim) [inline libs/contrib/Interfaces/Verified.idr]: Could we kindly remove the 'TODO' comments from the library code, or resolve them? I think it is not nice to have code that is commented out in the libraries.

File: libs/contrib/Interfaces/Verified.idr [hunk 2]
```diff
@@ -40,49 +69,118 @@ interface (Monad m, VerifiedApplicative m) => VerifiedMonad (m : Type -> Type) w
...
   total ringWithUnityIsUnityR : (r : a) -> Algebra.unity <.> r = r
 
---interface (VerifiedRingWithUnity a, Field a) => VerifiedField a where
---  total fieldInverseIsInverseL : (l : a) -> (notId : Not (l = neutral)) -> l <.> (inverseM l notId) = Algebra.unity
---  total fieldInverseIsInverseR : (r : a) -> (notId : Not (r = neutral)) -> (inverseM r notId) <.> r = Algebra.unity
+VerifiedRingWithUnity ZZ where
+  ringWithUnityIsUnityL = multOneRightNeutralZ
+  ringWithUnityIsUnityR = multOneLeftNeutralZ
 
+--interface (VerifiedRingWithUnity a, Field a) => VerifiedField a where
```
Comment (ahmadsalim) [inline libs/contrib/Interfaces/Verified.idr]: Could we also remove this commented out code, please?

File: libs/contrib/Interfaces/Verified.idr [hunk 3]
```diff
@@ -40,49 +69,118 @@ interface (Monad m, VerifiedApplicative m) => VerifiedMonad (m : Type -> Type) w
...
+[MultNatSemiV] VerifiedSemigroup Nat using MultNatSemi where
+  semigroupOpIsAssociative = multAssociative
+
+[PlusZZSemiV] VerifiedSemigroup ZZ using PlusZZSemi where
+  semigroupOpIsAssociative = plusAssociativeZ
+
+[MultZZSemiV] VerifiedSemigroup ZZ using MultZZSemi where
+  semigroupOpIsAssociative = multAssociativeZ
+
+-- TODO Verified versions of...
```
Comment (ahmadsalim) [inline libs/contrib/Interfaces/Verified.idr]: Same here.

File: libs/contrib/Interfaces/Verified.idr [hunk 4]
```diff
@@ -30,6 +40,25 @@ interface (Applicative f, VerifiedFunctor f) => VerifiedApplicative (f : Type ->
...
+  applicativeComposition Nothing (Just x) (Just y) = Refl
+  applicativeComposition (Just x) Nothing Nothing = Refl
+  applicativeComposition (Just x) Nothing (Just y) = Refl
+  applicativeComposition (Just x) (Just y) Nothing = Refl
+  applicativeComposition (Just x) (Just y) (Just z) = Refl
+  applicativeHomomorphism x g = Refl
+  applicativeInterchange x Nothing = Refl
+  applicativeInterchange x (Just y) = Refl
+
+-- TODO VerifiedApplicative instances for: List, Stream, Either e
```
Comment (ahmadsalim) [inline libs/contrib/Interfaces/Verified.idr]: Same here.

**Comment by atacratic** [comment]
@ahmadsalim Thanks for reviewing!  :smile:  Have pushed those cleanups.

**Comment by ahmadsalim** [comment]
@atacratic Sure 😄. I approved the changes now.

**Pull request merged**